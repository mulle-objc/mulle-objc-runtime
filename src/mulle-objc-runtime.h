//
//  mulle_objc.h
//  mulle-objc-runtime
//
//  Created by Nat! on 10.07.16.
//  Copyright (c) 2016 Nat! - Mulle kybernetiK.
//  Copyright (c) 2016 Codeon GmbH.
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#ifndef mulle_objc_runtime_h__
#define mulle_objc_runtime_h__

/*
 * If you want to include <mulle-objc-runtime/mulle-objc-runtime.h> you
 * should be compiling with an Objective-C compiler, that defines the
 * following defines properly. Do not use .c  use .m with mulle-objc-runtime.h
 */

// catch this early
#if ! defined( __MULLE_OBJC_TPS__) && ! defined( __MULLE_OBJC_NO_TPS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (or define either __MULLE_OBJC_TPS__ or __MULLE_OBJC_NO_TPS__)"
#endif
#if defined( __MULLE_OBJC_TPS__) && defined( __MULLE_OBJC_NO_TPS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (do not define both __MULLE_OBJC_TPS__ and __MULLE_OBJC_NO_TPS__)"
#endif
#if ! defined( __MULLE_OBJC_FCS__) && ! defined( __MULLE_OBJC_NO_FCS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (or define either __MULLE_OBJC_FCS__ or __MULLE_OBJC_NO_FCS__)"
#endif
#if defined( __MULLE_OBJC_FCS__) && defined( __MULLE_OBJC_NO_FCS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (do not define both __MULLE_OBJC_FCS__ and __MULLE_OBJC_NO_FCS__)"
#endif
#if ! defined( __MULLE_OBJC_TAO__) && ! defined( __MULLE_OBJC_NO_TAO__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (or define either __MULLE_OBJC_TAO__ or __MULLE_OBJC_NO_TAO__)"
#endif
#if defined( __MULLE_OBJC_TAO__) && defined( __MULLE_OBJC_NO_TAO__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (do not define both __MULLE_OBJC_TAO__ and __MULLE_OBJC_NO_TAO__)"
#endif

//
// these should have been defined by the compiler already, but for
// mulle-objc-runtime testing with gcc its convenient to add these if missing.
//
#ifndef __MULLE_OBJC_UNIVERSEID__
# define __MULLE_OBJC_UNIVERSEID__  0
#endif
#ifndef __MULLE_OBJC_UNIVERSENAME__
# define __MULLE_OBJC_UNIVERSENAME__  NULL
#endif

#if defined( __MULLE_OBJC_TPS__) && __MULLE_OBJC_UNIVERSEID__ != 0
# error "TPS can only be used in the default universe not in " #__MULLE_OBJC_UNIVERSENAME__
#endif


#include "include.h"

#include "mulle-metaabi.h"

#include "mulle-objc-atomicpointer.h"
#include "mulle-objc-builtin.h"
#include "mulle-objc-call.h"
#include "mulle-objc-class.h"
#include "mulle-objc-classpair.h"
#include "mulle-objc-class-convenience.h"
#include "mulle-objc-class-initialize.h"
#include "mulle-objc-class-lookup.h"
#include "mulle-objc-class-search.h"
#include "mulle-objc-class-struct.h"
#include "mulle-objc-csvdump.h"
#include "mulle-objc-fastclasstable.h"
#include "mulle-objc-fastmethodtable.h"
#include "mulle-objc-infraclass.h"
#include "mulle-objc-ivar.h"
#include "mulle-objc-ivarlist.h"
#include "mulle-objc-kvccache.h"
#include "mulle-objc-load.h"
#include "mulle-objc-metaclass.h"
#include "mulle-objc-method.h"
#include "mulle-objc-methodidconstants.h"
#include "mulle-objc-methodlist.h"
#include "mulle-objc-object.h"
#include "mulle-objc-objectheader.h"
#include "mulle-objc-object-convenience.h"
#include "mulle-objc-property.h"
#include "mulle-objc-propertylist.h"
#include "mulle-objc-retain-release.h"
#include "mulle-objc-signature.h"
#include "mulle-objc-super.h"
#include "mulle-objc-symbolizer.h"
#include "mulle-objc-taggedpointer.h"
#include "mulle-objc-try-catch-finally.h"
#include "mulle-objc-universe.h"
#include "mulle-objc-universe-class.h"
#include "mulle-objc-universe-exception.h"
#include "mulle-objc-universe-fail.h"
#include "mulle-objc-universe-global.h"
#include "mulle-objc-universe-struct.h"
#include "mulle-objc-uniqueid.h"
#include "mulle-objc-uniqueidarray.h"
#include "mulle-objc-version.h"
#include "mulle-objc-walktypes.h"



// add some functions to mulle-vararg for ObjC

#define mulle_vararg_count_ids( args, obj) \
   mulle_vararg_count_pointers( args, obj)
#define mulle_vararg_count_objects( args, obj) \
   mulle_vararg_count_pointers( args, obj)

#define mulle_vararg_next_id( args) \
   mulle_vararg_next_pointer( args, id)
#define mulle_vararg_next_object( args, type) \
   mulle_vararg_next_pointer( args, type)

#define mulle_vararg_push_id( args, value) \
   _mulle_vararg_push( args, id, value)
#define mulle_vararg_push_object( args, value) \
   mulle_vararg_push_id( args, value)

//
// add this before return of int main( argc, char **argc) when running
// named universes and leak checking
//
static inline void  mulle_objc_global_finish( void)
{
   struct _mulle_objc_universe   *universe;

   universe = mulle_objc_global_get_universe_inline( __MULLE_OBJC_UNIVERSEID__);
   if( universe && ! _mulle_objc_universe_is_default( universe) && universe->config.pedantic_exit)
      _mulle_objc_universe_release( universe);
}



#ifdef __has_include
# if __has_include( "_mulle-objc-runtime-versioncheck.h")
#  include "_mulle-objc-runtime-versioncheck.h"
# endif
#endif


#include <assert.h>
#include <stdarg.h>
#include <setjmp.h>

#endif

