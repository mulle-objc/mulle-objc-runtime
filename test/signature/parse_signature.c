//
//  main.c
//  test-runtime-2
//
//  Created by Nat! on 19/11/14.
//  Copyright (c) 2014 Mulle kybernetiK. All rights reserved.
//
#ifndef __MULLE_OBJC__
# define __MULLE_OBJC_NO_TPS__
# define __MULLE_OBJC_FCS__
# if defined( DEBUG) || ! defined( __OPTIMIZE__)
#  define __MULLE_OBJC_TAO__
# else
#  define __MULLE_OBJC_NO_TAO__
# endif
#endif



#include <mulle-objc-runtime/mulle-objc-runtime.h>

#include <stdio.h>


static char   *signatures[] =
{
   "v40@0:8*16Q24^v32",
   "#16@0:8",
   "@24@0:8@\"NSData\"16",
   "@\"MulleScionObject\"40@0:8@\"<MulleScionOutput>\"16@\"NSMutableDictionary\"24@\"<MulleScionDataSource>\"32",
   "@28@0:8i16@?<v@?>20",
   "i28@0:8q16f24",
   0
};


static char   *string_for_paramtype( enum mulle_metaabi_param type)
{
   switch( type)
   {
   case mulle_metaabi_param_error         : return( "error");
   case mulle_metaabi_param_void          : return( "void");
   case mulle_metaabi_param_void_pointer  : return( "void *");
   case mulle_metaabi_param_struct        : return( "struct {...} *");
   }
   return( "FAIL");
}


static int   test( char *s)
{
   unsigned int  i, n;
   char          *next;

   n = mulle_objc_signature_count_typeinfos( s);
   printf( "signature=\"%s\"\n", s);
   printf( "number of types = %u\n", n);
   printf( "metaabi param type = %s\n", string_for_paramtype( mulle_objc_signature_get_metaabiparamtype( s)));
   printf( "metaabi return type = %s\n", string_for_paramtype( mulle_objc_signature_get_metaabireturntype( s)));

   for( i = 0; i < n; i++)
   {
      next = mulle_objc_signature_next_type( s);
      printf( "#%d = %.*s\n", i, next ? (int) (next - s) : 255, s);
      s    = next;
   }

   if( ! s)
   {
      fprintf( stderr, "unexpected NULL in %s\n", __PRETTY_FUNCTION__);
      return( 1);
   }

   if( *s)
   {
      fprintf( stderr, "unexpected tail \"%s\" in %s\n", s, __PRETTY_FUNCTION__);
      return( 1);
   }
   return( 0);
}


static int   test_info( char *s)
{
   unsigned int  i, n;
   char          *next;
   struct mulle_objc_typeinfo   info;

   n = mulle_objc_signature_count_typeinfos( s);

   printf( "signature=\"%s\"\n", s);
   printf( "number of types = %u\n", n);

   for( i = 0; i < n; i++)
   {
      next = mulle_objc_signature_supply_typeinfo( s, NULL, &info);

      printf( "#%d type= %s\n", i, info.type ? info.type : "NULL");
      printf( "#%d pure_type_end= %s\n", i, info.pure_type_end ? info.pure_type_end : "NULL");
      printf( "#%d name= %s\n", i, info.name ? info.name : "NULL");
      printf( "#%d n_members= %ld\n", i, (long) info.n_members);
      printf( "#%d has_object= %d\n", i, info.has_object);
      printf( "#%d has_object= %d\n", i, info.has_object);

      s = next;
   }

   if( ! s)
   {
      fprintf( stderr, "unexpected NULL in %s\n", __PRETTY_FUNCTION__);
      return( 1);
   }
   if( *s)
   {
      fprintf( stderr, "unexpected tail \"%s\" in %s\n", s, __PRETTY_FUNCTION__);
      return( 1);
   }
   return( 0);
}



static int   test_while( char *s)
{
   unsigned int   i, n;
   char           *next;

   n = mulle_objc_signature_count_typeinfos( s);
   i = 0;
   while( (next = mulle_objc_signature_next_type( s)))
   {
      printf( "#%d = %.*s\n", i, next ? (int) (next - s) : 255, s);
      s = next;
      ++i;
   }

   if( i != n)
   {
      fprintf( stderr, "unexpected loop mismatch in %s\n", __PRETTY_FUNCTION__);
      return( 1);
   }

   return( 0);
}


static int   test_while_info( char *s)
{
   struct mulle_objc_typeinfo   info;
   unsigned int                 i, n;
   char                        *next;

   n = mulle_objc_signature_count_typeinfos( s);
   i = 0;

   while( (next = mulle_objc_signature_supply_typeinfo( s, NULL, &info)))
   {
      printf( "#%d: type=%s\n", i, info.type ? info.type : "NULL");
      s = next;
      ++i;
   }

   if( i != n)
   {
      fprintf( stderr, "unexpected loop mismatch in %s\n", __PRETTY_FUNCTION__);
      return( 1);
   }

   return( 0);
}


int   main( int argc, const char * argv[])
{
   char  **p;

   for( p = signatures; *p; p++)
   {
      if( test( *p))
         return( 1);
      putchar( '\n');

      if( test_while( *p))
         return( 1);
      putchar( '\n');

      if( test_info( *p))
         return( 1);
      putchar( '\n');

      if( test_while_info( *p))
         return( 1);
      putchar( '\n');
   }

   return 0;
}


