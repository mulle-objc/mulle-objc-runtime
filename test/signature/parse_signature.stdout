signature="v40@0:8*16Q24^v32"
number of types = 6
metaabi param type = struct {...} *
metaabi return type = void
#0 = v40
#1 = @0
#2 = :8
#3 = *16
#4 = Q24
#5 = ^v32

#0 = v40
#1 = @0
#2 = :8
#3 = *16
#4 = Q24
#5 = ^v32

signature="v40@0:8*16Q24^v32"
number of types = 6
#0 type                = v40@0:8*16Q24^v32
#0 pure_type_end       = 40@0:8*16Q24^v32
#0 name                = NULL
#0 n_members           = 0
#0 has_object          = 0
#0 has_retainable_type = 0
#1 type                = @0:8*16Q24^v32
#1 pure_type_end       = 0:8*16Q24^v32
#1 name                = NULL
#1 n_members           = 0
#1 has_object          = 1
#1 has_retainable_type = 1
#2 type                = :8*16Q24^v32
#2 pure_type_end       = 8*16Q24^v32
#2 name                = NULL
#2 n_members           = 0
#2 has_object          = 0
#2 has_retainable_type = 0
#3 type                = *16Q24^v32
#3 pure_type_end       = 16Q24^v32
#3 name                = NULL
#3 n_members           = 0
#3 has_object          = 0
#3 has_retainable_type = 1
#4 type                = Q24^v32
#4 pure_type_end       = 24^v32
#4 name                = NULL
#4 n_members           = 0
#4 has_object          = 0
#4 has_retainable_type = 0
#5 type                = ^v32
#5 pure_type_end       = 32
#5 name                = NULL
#5 n_members           = 0
#5 has_object          = 0
#5 has_retainable_type = 0

#0: type               = v40@0:8*16Q24^v32
#1: type               = @0:8*16Q24^v32
#2: type               = :8*16Q24^v32
#3: type               = *16Q24^v32
#4: type               = Q24^v32
#5: type               = ^v32

signature="#16@0:8"
number of types = 3
metaabi param type = void
metaabi return type = void *
#0 = #16
#1 = @0
#2 = :8

#0 = #16
#1 = @0
#2 = :8

signature="#16@0:8"
number of types = 3
#0 type                = #16@0:8
#0 pure_type_end       = 16@0:8
#0 name                = NULL
#0 n_members           = 0
#0 has_object          = 1
#0 has_retainable_type = 0
#1 type                = @0:8
#1 pure_type_end       = 0:8
#1 name                = NULL
#1 n_members           = 0
#1 has_object          = 1
#1 has_retainable_type = 1
#2 type                = :8
#2 pure_type_end       = 8
#2 name                = NULL
#2 n_members           = 0
#2 has_object          = 0
#2 has_retainable_type = 0

#0: type               = #16@0:8
#1: type               = @0:8
#2: type               = :8

signature="@24@0:8@"NSData"16"
number of types = 4
metaabi param type = void *
metaabi return type = void *
#0 = @24
#1 = @0
#2 = :8
#3 = @"NSData"16

#0 = @24
#1 = @0
#2 = :8
#3 = @"NSData"16

signature="@24@0:8@"NSData"16"
number of types = 4
#0 type                = @24@0:8@"NSData"16
#0 pure_type_end       = 24@0:8@"NSData"16
#0 name                = NULL
#0 n_members           = 0
#0 has_object          = 1
#0 has_retainable_type = 1
#1 type                = @0:8@"NSData"16
#1 pure_type_end       = 0:8@"NSData"16
#1 name                = NULL
#1 n_members           = 0
#1 has_object          = 1
#1 has_retainable_type = 1
#2 type                = :8@"NSData"16
#2 pure_type_end       = 8@"NSData"16
#2 name                = NULL
#2 n_members           = 0
#2 has_object          = 0
#2 has_retainable_type = 0
#3 type                = @"NSData"16
#3 pure_type_end       = "NSData"16
#3 name                = "NSData"16
#3 n_members           = 0
#3 has_object          = 1
#3 has_retainable_type = 1

#0: type               = @24@0:8@"NSData"16
#1: type               = @0:8@"NSData"16
#2: type               = :8@"NSData"16
#3: type               = @"NSData"16

signature="@"MulleScionObject"40@0:8@"<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32"
number of types = 6
metaabi param type = struct {...} *
metaabi return type = void *
#0 = @"MulleScionObject"40
#1 = @0
#2 = :8
#3 = @"<MulleScionOutput>"16
#4 = @"NSMutableDictionary"24
#5 = @"<MulleScionDataSource>"32

#0 = @"MulleScionObject"40
#1 = @0
#2 = :8
#3 = @"<MulleScionOutput>"16
#4 = @"NSMutableDictionary"24
#5 = @"<MulleScionDataSource>"32

signature="@"MulleScionObject"40@0:8@"<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32"
number of types = 6
#0 type                = @"MulleScionObject"40@0:8@"<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32
#0 pure_type_end       = "MulleScionObject"40@0:8@"<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32
#0 name                = "MulleScionObject"40@0:8@"<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32
#0 n_members           = 0
#0 has_object          = 1
#0 has_retainable_type = 1
#1 type                = @0:8@"<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32
#1 pure_type_end       = 0:8@"<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32
#1 name                = NULL
#1 n_members           = 0
#1 has_object          = 1
#1 has_retainable_type = 1
#2 type                = :8@"<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32
#2 pure_type_end       = 8@"<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32
#2 name                = NULL
#2 n_members           = 0
#2 has_object          = 0
#2 has_retainable_type = 0
#3 type                = @"<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32
#3 pure_type_end       = "<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32
#3 name                = "<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32
#3 n_members           = 0
#3 has_object          = 1
#3 has_retainable_type = 1
#4 type                = @"NSMutableDictionary"24@"<MulleScionDataSource>"32
#4 pure_type_end       = "NSMutableDictionary"24@"<MulleScionDataSource>"32
#4 name                = "NSMutableDictionary"24@"<MulleScionDataSource>"32
#4 n_members           = 0
#4 has_object          = 1
#4 has_retainable_type = 1
#5 type                = @"<MulleScionDataSource>"32
#5 pure_type_end       = "<MulleScionDataSource>"32
#5 name                = "<MulleScionDataSource>"32
#5 n_members           = 0
#5 has_object          = 1
#5 has_retainable_type = 1

#0: type               = @"MulleScionObject"40@0:8@"<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32
#1: type               = @0:8@"<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32
#2: type               = :8@"<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32
#3: type               = @"<MulleScionOutput>"16@"NSMutableDictionary"24@"<MulleScionDataSource>"32
#4: type               = @"NSMutableDictionary"24@"<MulleScionDataSource>"32
#5: type               = @"<MulleScionDataSource>"32

signature="@28@0:8i16@?<v@?>20"
number of types = 5
metaabi param type = struct {...} *
metaabi return type = void *
#0 = @28
#1 = @0
#2 = :8
#3 = i16
#4 = @?<v@?>20

#0 = @28
#1 = @0
#2 = :8
#3 = i16
#4 = @?<v@?>20

signature="@28@0:8i16@?<v@?>20"
number of types = 5
#0 type                = @28@0:8i16@?<v@?>20
#0 pure_type_end       = 28@0:8i16@?<v@?>20
#0 name                = NULL
#0 n_members           = 0
#0 has_object          = 1
#0 has_retainable_type = 1
#1 type                = @0:8i16@?<v@?>20
#1 pure_type_end       = 0:8i16@?<v@?>20
#1 name                = NULL
#1 n_members           = 0
#1 has_object          = 1
#1 has_retainable_type = 1
#2 type                = :8i16@?<v@?>20
#2 pure_type_end       = 8i16@?<v@?>20
#2 name                = NULL
#2 n_members           = 0
#2 has_object          = 0
#2 has_retainable_type = 0
#3 type                = i16@?<v@?>20
#3 pure_type_end       = 16@?<v@?>20
#3 name                = NULL
#3 n_members           = 0
#3 has_object          = 0
#3 has_retainable_type = 0
#4 type                = @?<v@?>20
#4 pure_type_end       = <v@?>20
#4 name                = NULL
#4 n_members           = 0
#4 has_object          = 0
#4 has_retainable_type = 0

#0: type               = @28@0:8i16@?<v@?>20
#1: type               = @0:8i16@?<v@?>20
#2: type               = :8i16@?<v@?>20
#3: type               = i16@?<v@?>20
#4: type               = @?<v@?>20

signature="i28@0:8q16f24"
number of types = 5
metaabi param type = struct {...} *
metaabi return type = void *
#0 = i28
#1 = @0
#2 = :8
#3 = q16
#4 = f24

#0 = i28
#1 = @0
#2 = :8
#3 = q16
#4 = f24

signature="i28@0:8q16f24"
number of types = 5
#0 type                = i28@0:8q16f24
#0 pure_type_end       = 28@0:8q16f24
#0 name                = NULL
#0 n_members           = 0
#0 has_object          = 0
#0 has_retainable_type = 0
#1 type                = @0:8q16f24
#1 pure_type_end       = 0:8q16f24
#1 name                = NULL
#1 n_members           = 0
#1 has_object          = 1
#1 has_retainable_type = 1
#2 type                = :8q16f24
#2 pure_type_end       = 8q16f24
#2 name                = NULL
#2 n_members           = 0
#2 has_object          = 0
#2 has_retainable_type = 0
#3 type                = q16f24
#3 pure_type_end       = 16f24
#3 name                = NULL
#3 n_members           = 0
#3 has_object          = 0
#3 has_retainable_type = 0
#4 type                = f24
#4 pure_type_end       = 24
#4 name                = NULL
#4 n_members           = 0
#4 has_object          = 0
#4 has_retainable_type = 0

#0: type               = i28@0:8q16f24
#1: type               = @0:8q16f24
#2: type               = :8q16f24
#3: type               = q16f24
#4: type               = f24

